cdap-ui/app/directives/metric-picker/metric-picker.js

```diff
@@ -109,45 +145,54 @@ angular.module(PKG.name + '.commons')
         elem.find('button').on('blur', onBlurHandler);
         elem.find('input').on('blur', onBlurHandler);
 
-        scope.$watchCollection('metric', function (newVal, oldVal) {
+        var metricChanged = function (newVal, oldVal) {
           ngModel.$validate();
 
           if(newVal.type !== oldVal.type) {
             scope.metric.context = '';
-            scope.metric.name = '';
+            scope.metric.resetNames();
             fetchAhead();
             return;
           }
 
           if(newVal.context !== oldVal.context) {
-            scope.metric.name = '';
+            scope.metric.resetNames();
             fetchAhead();
+            return;
           }
 
-          if(newVal.context && newVal.name) {
-            if (newVal.name === 'Add All') {
+          if(newVal.context && newVal.names) {
+            var isAddAll = false;
+            for (var i = 0; i < newVal.names.length; i++) {
+              if (newVal.names[i].name === 'Add All') {
+                isAddAll = true;
+              }
+            }
+            if (isAddAll) {
               ngModel.$setViewValue({
                 addAll: true,
-                allMetrics: scope.available.names.splice(1), // Remove 'Add All' option
+                allMetrics: scope.available.names.slice(1), // Remove 'Add All' option
                 context: getBaseContext() + '.' + newVal.context,
-                name: newVal.name
+                names: newVal.getNames(),
+                name: newVal.getName()
               });
               return;
             } else {
               ngModel.$setViewValue({
                 context: getBaseContext() + '.' + newVal.context,
-                name: newVal.name
+                names: newVal.getNames(),
+                name: newVal.getName()
               });
             }
-          }
-          else {
+          } else {
             if(ngModel.$dirty) {
               ngModel.$setViewValue(null);
             }
 
           }
 
-        });
+        }
+        scope.$watch('metric', metricChanged, true);
```

Comment:
So when we do ```$scope.$watch( 'something', handler , true)``` then it makes a deep-copy check.

In our case we have a object like this
```
{
  context: '',
  metric: [ ... ],
  type: ''
}
```
when we do ```$watch(something, handler)``` it will trigger a change event when we modify an array item in metric (say in our case name). So this needs the third argument to be true but on a longer run we need to figure out how to avoid that. Not necessarily in this PR.